### ISSUE TITLE
`plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)`

### ISSUE THREAD

**Issue by ShockwaveNN** [rubygems/rubygems#5289]
Hi, I'm having a problem, I've tried to search it but seems it's rather hard to find any relative info to my problem, so please point out to correctissue if it's already exists

I maintain several gems, I push it to RubyGems and Github packages and I got similar-looking Rakefile:
```
# frozen_string_literal: true

require 'bundler/gem_tasks'

desc 'Release gem '
task :release_github_rubygems do
  Rake::Task['release'].invoke
  gem_name = "pkg/name-version.gem"
  sh('gem push --key github '\
     '--host https://rubygems.pkg.github.com/team '\
     "#{gem_name}")
end
```

Everything worked fine for a long time but with Ruby 3.1 I got those logs:
```
rake release_github_rubygems
gem_name 0.4.0 built to pkg/gem_name-0.4.0.gem.
Tagged v0.4.0.
Pushed git commits and release tag.
Error loading RubyGems plugin "/home/lobashov/.rvm/gems/ruby-3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Error loading RubyGems plugin "/home/lobashov/.rvm/rubies/ruby-3.1.0/lib/ruby/gems/3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Pushing gem to https://rubygems.org...
You have enabled multi-factor authentication. Please enter OTP code.
Code:   123456
Successfully registered gem: gem_name (0.4.0)
Pushed gem_name 0.4.0 to rubygems.org
gem push --key github --host https://rubygems.pkg.github.com/team pkg/gem_name-0.4.0.gem
Error loading RubyGems plugin "/home/lobashov/.rvm/gems/ruby-3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Error loading RubyGems plugin "/home/lobashov/.rvm/rubies/ruby-3.1.0/lib/ruby/gems/3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Pushing gem to https://rubygems.pkg.github.com/team...
Successfully registered gem: gem_name (0.4.0)
```

Stranges thing - release worked OK, I got new version of gem on both rubygems and github Packages, but I don't like error messages in logs

Here are my current environment details:

```
RubyGems Environment:
  - RUBYGEMS VERSION: 3.3.4
  - RUBY VERSION: 3.1.0 (2021-12-25 patchlevel 0) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/lobashov/.rvm/gems/ruby-3.1.0
  - USER INSTALLATION DIRECTORY: /home/lobashov/.gem/ruby/3.1.0
  - RUBY EXECUTABLE: /home/lobashov/.rvm/rubies/ruby-3.1.0/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /home/lobashov/.rvm/gems/ruby-3.1.0/bin
  - SPEC CACHE DIRECTORY: /home/lobashov/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/lobashov/.rvm/rubies/ruby-3.1.0/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /home/lobashov/.rvm/gems/ruby-3.1.0
     - /home/lobashov/.rvm/rubies/ruby-3.1.0/lib/ruby/gems/3.1.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/lobashov/.rvm/gems/ruby-3.1.0/bin
     - /home/lobashov/.rvm/gems/ruby-3.1.0@global/bin
     - /home/lobashov/.rvm/rubies/ruby-3.1.0/bin
     - /home/lobashov/.rvm/bin
     - /home/lobashov/.local/bin
     - /home/lobashov/bin
     - /home/lobashov/yandex-cloud/bin
     - /home/lobashov/.local/bin
     - /home/lobashov/bin
     - /usr/local/sbin
     - /usr/local/bin
     - /usr/sbin
     - /usr/bin
     - /sbin
     - /bin
     - /usr/games
     - /usr/local/games
     - /snap/bin
     - /home/lobashov/.dotnet/tools
     - /home/lobashov/bin
     - /home/lobashov/.dotnet/tools
     - /home/lobashov/bin
     - /home/lobashov/bin
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by deivid-rodriguez** [comment]
Thanks for the report @ShockwaveNN. Does the following print an existing path or an error?

```
ls /home/lobashov/.rvm/gems/ruby-3.1.0/gems/gem-release-$(ruby -e 'print Gem.latest_spec_for("gem-release").version.to_s')/lib/gem/release.rb
```

**Comment by ShockwaveNN** [comment]
@deivid-rodriguez Thanks for quick reply

```
# ls /home/lobashov/.rvm/gems/ruby-3.1.0/gems/gem-release-$(ruby -e 'print Gem.latest_spec_for("gem-release").version.to_s')/lib/gem/release.rb

/home/lobashov/.rvm/gems/ruby-3.1.0/gems/gem-release-2.2.2/lib/gem/release.rb
```
Seems that it's the latest version: https://rubygems.org/gems/gem-release/versions/2.2.2

**Comment by deivid-rodriguez** [comment]
Ok, and is this gem properly installed, i.e., does `gem list` show it?

**Comment by ShockwaveNN** [comment]
Yep

```
# gem list | grep release
gem-release (2.2.2)
```

Stangest thing - that release process IS FINE, so all I need to be released - released ok, but with errors in logs

**Comment by deivid-rodriguez** [comment]
Yeah, errors when loading RubyGems plugins are only warned, but not fatal:

https://github.com/rubygems/rubygems/blob/ecb6c4abb399fcd740f8a4165dc068cff77ebc1a/lib/rubygems.rb#L1029-L1034

Question is why this require is failing:

https://github.com/svenfuchs/gem-release/blob/master/lib/rubygems_plugin.rb#L1

**Comment by ShockwaveNN** [comment]
@deivid-rodriguez Problem is that I cannot correctly simulate this error without making actual release of some gems (and I don't want it as you understand)

Maybe you can provide some command that will raise similar errors without making releases and after that, I'll try to reproduce same error in some Dockerfile?

**Comment by deivid-rodriguez** [comment]
Yeah, I understand, but I'm not sure. Technically RubyGems plugins are loaded on every Ruby invocation, unless `--disable-gems` flags is given to Ruby. I have no idea what makes your case special.

Can you try duplicating your task but running a different `gem` command, like say `gem list`, and see if the error still reproduces?

**Comment by ShockwaveNN** [comment]
Got it thanks, actually you're right

I was able to reproduce it with this Rakefile

```
# frozen_string_literal: true

require 'bundler/gem_tasks'
require 'rspec/core/rake_task'

RSpec::Core::RakeTask.new(:spec)

task default: :spec

desc 'Release gem '
task :release_github_rubygems do
  `gem list`
end
```

output is:
```
$ rake release_github_rubygems
Error loading RubyGems plugin "/home/lobashov/.rvm/gems/ruby-3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Error loading RubyGems plugin "/home/lobashov/.rvm/rubies/ruby-3.1.0/lib/ruby/gems/3.1.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
```
I'll try to figure your why and make a Dockerfile to reproduce it

**Comment by deivid-rodriguez** [comment]
Great, I tried something similar and couldn't reproduce the issue, so there must be something going on in your environment. It could be `RVM` related since I use `Rbenv`.

Other culprit I could imagine is if you had some `RUBYOPT=--disable-gems` environment variable hanging around somewhere.

**Comment by ShockwaveNN** [comment]
So I was able to reproduce on my system with simplest project

`Gemfile`
```
# frozen_string_literal: true

source 'https://rubygems.org'

gemspec
```

`Rakefile`
```
# frozen_string_literal: true

require 'bundler/gem_tasks'

desc 'Release gem'
task :release_github_rubygems do
  `gem list`
end
```

`test.gemspec`:
```
# frozen_string_literal: true

Gem::Specification.new do |s|
  s.name = 'test'
  s.version = '0.0.1'
  s.authors = ['test']
  s.summary = 'test'
  s.add_development_dependency('rake', '~> 13')
end
```

And noticed that `s.add_development_dependency('rake', '~> 13')` is crucial, without this line there is no errors

But currently no luck in making Dockerfile

**Comment by ShockwaveNN** [comment]
And RVM based build has no errors either

```
FROM ubuntu:20.04

RUN apt-get -y update && apt-get -y install curl \
                                            gpg 
RUN gpg --keyserver keyserver.ubuntu.com --recv-key 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
RUN curl -sSL https://get.rvm.io | bash -s stable
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && rvm install 3.1.0"

RUN mkdir -pv /test-gem
WORKDIR /test-gem
RUN echo "# frozen_string_literal: true\nsource 'https://rubygems.org'\ngemspec" > Gemfile
RUN echo "# frozen_string_literal: true\nGem::Specification.new do |s|\ns.name = 'test'\ns.version = '0.0.1'\ns.authors = ['test']\ns.summary = 'test'\ns.add_development_dependency('rake', '~> 13')\nend" > test.gemspec
RUN echo "# frozen_string_literal: true\nrequire 'bundler/gem_tasks'\ndesc 'Release gem'\ntask :release_github_rubygems do\n\`gem list\`\nend" > Rakefile
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && bundler install"
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && rake release_github_rubygems"
```

Seems this has something to do with my machine and I have no idea that
I've see no changes at `RUBYOPT` anywhere

@deivid-rodriguez I think we can close this issue or leave this issue open - maybe someone has similar problems. It's up to you

**Comment by deivid-rodriguez** [comment]
You need to add installing the `gem-release` gem to those Dockerfiles, right? Otherwise it's not going to reproduce.

**Comment by ShockwaveNN** [comment]
@deivid-rodriguez 

No luck for both variant:
```
FROM ruby:3.1.0

RUN mkdir -pv /test-gem
WORKDIR /test-gem
RUN echo "# frozen_string_literal: true\nsource 'https://rubygems.org'\ngemspec" > Gemfile
RUN echo "# frozen_string_literal: true\nGem::Specification.new do |s|\ns.name = 'test'\ns.version = '0.0.1'\ns.authors = ['test']\ns.summary = 'test'\ns.add_development_dependency('rake', '~> 13')\nend" > test.gemspec
RUN echo "# frozen_string_literal: true\nrequire 'bundler/gem_tasks'\ndesc 'Release gem'\ntask :release_github_rubygems do\n\`gem list\`\nend" > Rakefile
RUN bundler install
RUN gem install gem-release
RUN rake release_github_rubygems
```

and

```
FROM ubuntu:20.04

RUN apt-get -y update && apt-get -y install curl \
                                            gpg 
RUN gpg --keyserver keyserver.ubuntu.com --recv-key 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
RUN curl -sSL https://get.rvm.io | bash -s stable
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && rvm install 3.1.0"

RUN mkdir -pv /test-gem
WORKDIR /test-gem
RUN echo "# frozen_string_literal: true\nsource 'https://rubygems.org'\ngemspec" > Gemfile
RUN echo "# frozen_string_literal: true\nGem::Specification.new do |s|\ns.name = 'test'\ns.version = '0.0.1'\ns.authors = ['test']\ns.summary = 'test'\ns.add_development_dependency('rake', '~> 13')\nend" > test.gemspec
RUN echo "# frozen_string_literal: true\nrequire 'bundler/gem_tasks'\ndesc 'Release gem'\ntask :release_github_rubygems do\n\`gem list\`\nend" > Rakefile
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && bundler install && gem install gem-release"
RUN /bin/bash -l -c ". /etc/profile.d/rvm.sh && rake release_github_rubygems"
```

Run fine

**Comment by ShockwaveNN** [comment]
I've remove RVM and reinstalled it anew and seems problem is gone

So thanks for your help, I think it was some specific for my configuration

**Comment by deivid-rodriguez** [comment]
Glad you sorted it out!

**Comment by ShockwaveNN** [comment]
And after that i've moved to `rbenv`

Maybe it's really good idea to let go of `rvm` )))

**Comment by jrochkind** [comment]
I'm getting this error on running `bundle exec` (anything I think?) in a gem source dir. Although the gem release seems to have happened succesfully anyway? 

But there is something weird going on. I use chruby (not rvm) to manage ruby versions. bundler 2.3.14, rake 13.0.6, rubygems 3.2.33. 

Oddly `bundle exec gem --version` also prints out the error. But not just any `bundle exec` will. 

So there is something weird going on.

**Comment by deivid-rodriguez** [comment]
You can start by upgrading all the things and see if it still reproduces!

**Comment by jrochkind** [comment]
So I'm not ready to update ruby (and after all ruby 3.0 is still supported, yes?).  Looks like rake is already latest. But let's look at the others...

after a `gem update --system` and a `gem install bundler`, and `bundle update --bundler`, I am at:  

ruby 3.0.4, bundler 2.3.20, rubygems 3.3.20, rake 13.0.6.

Somehow the problem has gotten even worse, now it reports the error twice?

```
$ bundle exec gem --version
Error loading RubyGems plugin "/Users/jrochkind/.gem/ruby/3.0.4/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
Error loading RubyGems plugin "/Users/jrochkind/.rubies/ruby-3.0.4/lib/ruby/gems/3.0.0/plugins/gem-release_plugin.rb": cannot load such file -- gem/release (LoadError)
3.3.20
```

**Comment by jrochkind** [comment]
The contents of those files I see are a require relative, pointing to... this one seems to be there too:

```
 $ ls /Users/jrochkind/.gem/ruby/3.0.4/gems/gem-release-2.2.2/lib/rubygems_plugin.rb
/Users/jrochkind/.gem/ruby/3.0.4/gems/gem-release-2.2.2/lib/rubygems_plugin.rb
```

**Comment by jrochkind** [comment]
Looking back at the discussion... I'm confused about what the `gem-release` gem _is_.  I have not intentionally installed it at any point, as far as I know, I am not familiar with it. 

I have been using the `bundle exec rake release` for years, with `require 'bundler/gem_tasks'` in my Rakefile.  I am not sure how these things are related, or what's going on. Before reading this thread, I was not aware of a gem called `gem-release`.

**Comment by jrochkind** [comment]
I tried `gem uninstall gem-release` followed by `gem install gem-release`. No change, still the same output with two errors on `bundle exec gem --version`.

**Comment by N0xFF** [comment]
Adding `gem "gem-release"` to `Gemfile` should solve your problem.

**Comment by uhuntu** [comment]
Yeah, you are smart.

**Comment by pboling** [comment]
Lots of plugins use the gem `gem-release`, which isn't part of bundler / rubygems, but appears to conflict in load paths, resulting in this error.

If you use a global Ruby for multiple projects (as you often do with any type of ruby version manager), some gems or apps will depend on the `gem-release`, and thus it will be installed and within reach of `gem`'s load paths.

Then when you do a canonical `bundle exec rake release` using the standard gem tasks, in a project that does not depend on `gem-release`, but within a Ruby environment that has it installed, it will mistakenly hit that load path.

You can see that is exactly what happened in the paths @jrochkind had:
```
/Users/jrochkind/.gem/ruby/3.0.4/gems/gem-release-2.2.2/lib/rubygems_plugin.rb
```

I just had the same issue.  I am not sure which library depended on it, but the temporary solution (until said library installs it again!) is to `gem uninstall gem-release`.

This is a bug, though perhaps it should be [re-opened](https://github.com/svenfuchs/gem-release/issues/105) on the `gem-release` tracker.

**Comment by PikachuEXE** [comment]
I opened a PR https://github.com/svenfuchs/gem-release/pull/111 and pushed a test version

If you can reproduce the error you can try the test version by
- `gem uninstall gem-release`
- `gem install gem-release -v '2.2.3.alpha1'`

I need 1-2 people to confirm it works before I merge and release it

**Comment by PikachuEXE** [comment]
The fix is updated if you have already tested `2.2.3.alpha1`
Please see PR for details to learn about the update & retest

**Comment by PikachuEXE** [comment]
Fix released in https://rubygems.org/gems/gem-release/versions/2.2.3
